### DISCUSSION TITLE
<vector>

### DISCUSSION THREAD

**Discussion by forthnutter** [factor/factor#2400] [General]
I notice something strange about `<vector>` word, the description is "Creates a new vector that can hold n elements before resizing".
I feed it 8 elements eg. 8 <vector>  the result is always V{ } an empty vector. I would expect to get back a vector with 8 elements all zero.
It looks like the definition is always zeroing the size of the vector tuple when using boa word
`: <vector> ( n -- vector ) 0 <array> 0 vector boa ; inline`

would it have been better to 
`: <vector> ( n -- vector ) 0 <array> >vector ; inline`

**Comment by alex-ilin** [comment]
You have to think about it as of capacity, not element count. If you want vector to contain prefilled elements, use something like n <array> >vector.

-------- Original Message --------

**Reply by forthnutter** [reply]
sorry some word  don't come out way you write them on this page.
Yes I could just write `8 0 <array> >vector`  so what is the purpose of `<vector>` if it is not intended to be initialized with a number of elements.
If I wanted an empty vector I could just do V{ } clone .

**Reply by alex-ilin** [reply]
What if you wanted an empty vector and knew that it'll grow to exactly (or maximum) 16 elements when you fill it up?
This is the reason you may set the initial capacity and avoid reallocating the memory after each individual element is added.
It saves CPU time and memory reallocations if you know that you'll need at least some capacity before you start adding stuff.

**Reply by alex-ilin** [reply]
Also if you create a vector with capacity 8 and give its memory address to some C function, you can add up to 8 elements and know that the address is still valid, i.e. the vector was not reallocated in memory: resizing often means allocating to a new place and copying the contents.

That's why documentation says "**can** hold n elements before resizing", it's an important feature. But it doesn't say that it "**does** hold n elements".

**Reply by forthnutter** [reply]
Thanks for that now I understand.
I see we do have two words that has helped "length and capacity ".
`8 <vector> `
`V{ } length`
`0`
`8 <vector>`
`V{ } capacity`
`8`
My thinking was all wrong.

**Comment by jonenst** [comment]
On github you have to quote with backticks : `<vector>`

It's pretty standard to create empty containers with an initial capacity,
for example in java
https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#ArrayList-int-

So I don't think we should change anything here.

Le jeu. 17 déc. 2020 à 13:34, Joseph Moschini <notifications@github.com> a
écrit :
